ocaml/xapi/xapi_psr.ml

```diff
@@ -186,24 +187,134 @@ functor
 
 let perm = 0o640
 
-let cleanup_internal ~additional_files_to_remove () =
-  if List.length !Xapi_globs.pool_secrets <= 1 then
-    D.info "xapi_psr.ml:cleanup_internal: already cleaned up"
-  else
-    (* if we don't remove pool secret backups, they will be loaded into memory when xapi starts *)
-    let files_to_remove =
-      List.append additional_files_to_remove
-        [old_pool_secret_backup_path; new_pool_secret_backup_path]
+(* we include some sanity checks for the following invariants:
+     - only the master should have a checkpoint
+     - either (a) all the master psr state exists (=> resuming a failed psr)
+       or     (b) none of the state exists (=> starting a new psr)
+     - existing pool secret backups are consistent with requested pool secret
+       changes
+     - the runtime state, i.e. Xapi_globs.pool_secrets (although this check
+       is not done here, but at each stage separately)
+   these invariants could be broken if for example a psr fails, the master
+   changes, and then a new psr starts on the new master
+*)
+module Assert : sig
+  val backups_match : old_ps:SecretString.t -> new_ps:SecretString.t -> unit
+
+  val no_backups : unit -> unit
+
+  val no_checkpoint : unit -> unit
+
+  val master_state_valid : unit -> unit
+end = struct
+  let do_backups_exist () =
+    let does_old_backup_exist = Sys.file_exists old_pool_secret_backup_path in
+    let does_new_backup_exist = Sys.file_exists new_pool_secret_backup_path in
+    match (does_old_backup_exist, does_new_backup_exist) with
+    | true, true ->
+        true
+    | false, false ->
+        false
+    | false, true | true, false ->
+        raise
+          Api_errors.(
+            Server_error
+              (internal_error, ["do_backups_exist: invalid backup state"]))
+
+  let does_checkpoint_exist () = Sys.file_exists checkpoint_path
+
+  let backups_match ~old_ps ~new_ps =
```

Comment:
A comment on style: this mixes logic and logging. I think it would be clearer to have a predicate for the matching and do the logging where it is used:

```
let backups_match = 
  let old_backup, new_backup = read_backups () in
  SecretString.equal old_backup old_ps
           && SecretString.equal new_backup new_ps
...
if not  backups_match then ..
```

Reply:
To me it seems like it's also mixing the result of reading the files and the result of the match.
This could be encoded as `(boolean, string) result` to completely separate business logic from error-treating logic.

```
match do_backups_match with
| Ok true -> ()
| _ -> raise Api_errors.(...)
```
The false and the failure to read ca be separated here as well to log them if need be.

Reply:
This is combination with your comment below regarding `read_backups` has tidied this up

Reply:
I don't gain much by returning a `result`, since we do always expect the backups to exist whenever we ask for them (see https://github.com/xapi-project/xen-api/pull/4123#discussion_r434637042)